### ISSUE TITLE
Parsing 2.220446049250313e-16 gives wrong result

### ISSUE THREAD

**Issue by segmtfault** [miloyip/rapidjson#340]
When parsing the following json document (with doc.Parse<rapidjson::kParseFullPrecisionFlag> ) : 

{
	"foo": 2.220446049250313e-16
} 

ParseNumber will call handler.Double with a wrong value, actually 1.1102230246251565e-016 which is half of the correct value.

What I have been able to dig so far:
ParseNumber will call internal::StrtodFullPrecision with the following values: d=2220446049250313.0, p=-31, decimal="2220446049250313", length=16, decimalPosition=1, exp=-16 , which all look correct.

In StrtodFullPrecision, the call to StrtodFast fails, so we fall down to StrtodDiyFp which returns the incorrect result.
I don't yet follow all the logic in StrtodDiyFp, but we end up with this values in rounded just before the call to rounded.ToDouble() : f = 9007199254740992, e=-105.
The problem is that f = 2^53 exactly, so  f & kDpSignificandMask (in ToDouble) results in 0, and we end up with the wrong double.

I don't know yet how/why we end up with f = 2^53, e=-105 instead of f = 2^52, e=-104.
I'll keep looking but hopefully you will figure it out before me :)

**Comment by segmtfault** [comment]
Update: this impacts other negative powers of 2. Running all negative powers of 2, from -1 to -100, yields the following error cases (where "foo_nn" has the value 2^-nn):
{
    "foo_27": 7.450580596923828e-9,
    "foo_28": 3.725290298461914e-9,
    "foo_29": 1.862645149230957e-9,
    "foo_40": 9.094947017729282e-13,
    "foo_41": 4.547473508864641e-13,
    "foo_51": 4.440892098500626e-16,
    "foo_52": 2.220446049250313e-16,
    "foo_58": 3.469446951953614e-18,
    "foo_59": 1.734723475976807e-18,
    "foo_64": 5.421010862427522e-20,
    "foo_65": 2.710505431213761e-20,
    "foo_66": 1.3552527156068805e-20,
    "foo_70": 8.470329472543003e-22,
    "foo_74": 5.293955920339377e-23,
    "foo_92": 2.0194839173657902e-28,
    "foo_93": 1.0097419586828951e-28,
}

This all result in rounded.f = 2^53

**Comment by miloyip** [comment]
The bug is reproduced. I am looking into it.
Thank you very much for your report.

**Comment by miloyip** [comment]
It is due to the last step in `StrtodDiyFp()`. After rounding, if it is a negative power of 2, it will overflow after adding 1 to the LSB of mantissa. And then the mantissa is masked for generating a `double`, the extra bit was masked out.

So the fix is to detect this situation and right-shift mantissa by 1 and increases exponent by 1.

Really thank for reporting this. I am curious how you find out this bug.

**Comment by segmtfault** [comment]
Thank you for the very quick fix!
How I found out: we have an optional "output to json" module in our Maya export pipeline, and this value would show up in some of our Maya files (as the Y of a transform's translate component, IIRC).
It showed up as a mismatch when I compared the output of our "export Maya to our binary format" and "export Maya to json, then convert the json to our binary format" paths.
